### DISCUSSION TITLE
After updating ruby-3.2.2 and using pundit-2.1 version, Getting error:  Invalid constructor<> is called

### DISCUSSION THREAD

**Discussion by pragya-sriharsh** [varvet/pundit#777] [Q & A] [answered]
Recently i tried with ruby 3.0.0 &  ruby-3.2.2 and using pundit-2.1 version. It gives me an error on this line

`authorize :home`

<img width="1084" alt="Screenshot 2023-06-08 at 3 21 39 PM" src="https://github.com/varvet/pundit/assets/118921874/a87a01b9-269e-44ae-badd-8d5c9ad8f1c2">

After refreshing two times, it is working fine and didn't get error again. After clearing browser cache then error comes again. It's weird.

Any comment on this?

I have tried with the new version of pundit - 2.3.0 and got same error. You have changed the authorize , policy & policy_scope! method in the 2.3.0 version. Can you guide me what will be write in place of `authorize :home` ? because i am getting the same error: Invalid constructor.

**Comment by Burgestrand** [comment]
I tried, but I can't find a way to reproduce your error. You need to provide some more information. A detailed backtrace would provide more clues, since the `InvalidConstructorError` is raised from an underlying `ArgumentError`. The backtrace of the [cause](https://rubyapi.org/3.2/o/exception#method-i-cause) might show us what's actually happening.

Absolute best would be if you could provide a small example we can run ourselves that shows your error.

From the context you've provided, I'm thinking:
* The code called by `authorize` that raises this error is here. From this code I would expect the policy class name to be present in the error. https://github.com/varvet/pundit/blob/16554a5fe814153d05cec133705bd583709a4124/lib/pundit.rb#L152-L157
* I would expect the error to at the very least be `Invalid #<HomePolicy> constructor is called`, but your policy class doesn't render its' name as a string so we can't tell which policy is actually being used.
* `nil.to_s` does render an empty string, but considering `PolicyFinder#policy!` didn't raise an error it's unlikely that `policy` is `nil` here.
* it's also unlikely that `policy` is `nil` because `nil.new(user, pundit_model(record))` doesn't raise an `ArgumentError`, it raises a `NoMethodError`

**Comment by pragya-sriharsh** [comment]
Previously i was using ruby version 2.7.6, So everything was working fine. After updating the ruby version 3.0.0, i am facing this issue. I am trying to find out what is the root cause.  

In terminal, i can see only few lines of backtrace(see screenshot)

<img width="1344" alt="Screenshot 2023-06-14 at 2 14 00 PM" src="https://github.com/varvet/pundit/assets/118921874/b5f6b975-b39a-450d-a842-60ea3a7cbcb0">


May be you can get the following error by configuring these settings:

ruby3.0.0 version
rails 6.1.7.3 version
pundit 2.1 version

Create HomeController and Create HomePolicy

```
class HomeController < ApplicationController
  def index
    authorize :home
  end
end
```

```
class HomePolicy < ApplicationPolicy
  def index?
    true
  end
end
```

If we can modify in other way to make it work then i am open. Let me know if you want anything or we can connect.

**Comment by Burgestrand** [comment]
I'm sorry, I don't think I can help you with that information.

I would advise you to figure out why the `Invalid #<> constructor is called` has an empty class name in it. Something is amiss there.

The Pundit source code is relatively small. There's not a lot to dig through.

I tried to replicate the code you sent in a full example, but it runs just fine with no errors. It took too long trying to figure out how to bootstrap Rails in this tiny example.

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"

  ruby "3.2.2"

  gem "pundit", "2.1.0"
end

class ApplicationController
  include Pundit

  def current_user = "Not actual User"
end

class HomeController < ApplicationController
  def index
    authorize :home

    "Works fine!"
  end

  private

  # Rails sets this. Pundit uses it to find the query method.
  def action_name = :index
end

class ApplicationPolicy
  def initialize(user, record)
    @user = user
    @record = record
  end
end

class HomePolicy < ApplicationPolicy
  def index? = true
end

controller = HomeController.new
puts controller.index # => "Works fine!"
```

**Comment by pragya-sriharsh** [comment, marked as answer]
Its a misleading error. Problem was in delegation method in ruby 3.0. Thanks for your support. Appreciate for that!
